comms/dht/src/crypt.rs

```diff
@@ -64,14 +64,10 @@ pub struct AuthenticatedCipherKey(chacha20poly1305::Key);
 const MESSAGE_BASE_LENGTH: usize = 6000;
 
 /// Generates a Diffie-Hellman secret `kx.G` as a `chacha20::Key` given secret scalar `k` and public key `P = x.G`.
-pub fn generate_ecdh_secret(secret_key: &CommsSecretKey, public_key: &CommsPublicKey) -> [u8; 32] {
-    // TODO: PK will still leave the secret in released memory. Implementing Zerioze on RistrettoPublicKey is not
-    //       currently possible because (Compressed)RistrettoPoint does not implement it.
+pub fn generate_ecdh_secret(secret_key: &CommsSecretKey, public_key: &CommsPublicKey) -> Zeroizing<Vec<u8>> {
     let k = CommsPublicKey::shared_secret(secret_key, public_key);
-    let mut output = [0u8; 32];
 
-    output.copy_from_slice(k.as_bytes());
-    output
+    Zeroizing::new(k.to_vec())
```

Comment:
Great, I think that the best choice we have is to heap-allocate the secret as you have, as stack types (like `[u8; 32]`) have a leaking problem as they are moved. [As stated by zeroize](https://docs.rs/zeroize/0.7.0/zeroize/index.html#stackheap-zeroing-notes).

I think we'll have to wait for  https://github.com/tari-project/tari-crypto/pull/137 to be merged so that we can call `k.zeroize()` before returning. Otherwise, `k` would still leak?

Reply:
Yep, merging should be considered blocked until the `tari-crypto` PR is merged.

We shouldn't need to explicitly call `k.zeroize()` anymore, since a recent change to the `tari-crypto` PR returns the shared secret in a `Zeroize` wrapper as well.

Reply:
Another thought I had was to build a more comprehensive generic shared secret type that zeroizes on drop and provides its own byte representation in `tari-crypto`. This would allow things like KDFs to do things like `shared_secret.as_bytes()` and mean we wouldn't need to handle the separate `Zeroizing` wrapper around the byte array, increasing safety.

Reply:
Ah yup that's right, missed that the shard_secret function now returns the wrapper. 
I agree it's a bit leaky and because it is that it is now part of the public API of tari-crypto, in most cases, users would have to include zerioze explicitly as a dependency just to use tari-crypto, or you would have to re-export it from tari-crypto.
I don't think that is the end of the world, but a newtype feels more ergonomic and safer here.

Reply:
Separate PRs, or part of this and the `tari-crypto` PRs?

Reply:
Up to you, this can go in as is. If you'd like to continue working in this PR mark as draft and it wont auto-merge :)

Reply:
Working on adding this now. It'll be a generic `DiffieHellmanSharedSecret<P>` for arbitrary `Zeroize` public key types, will zeroize on drop, and only provides `as_bytes(&self) -> &[u8]` as visibility into the shared secret. So the caller would need to go out of its way to leave a non-zeroizing copy in memory.